server/hotrod/src/main/java/org/infinispan/server/hotrod/CacheRequestProcessor.java

```diff
@@ -0,0 +1,627 @@
+import org.infinispan.persistence.manager.PersistenceManager;
+import org.infinispan.security.impl.SecureCacheImpl;
+import org.infinispan.server.hotrod.iteration.IterableIterationResult;
+import org.infinispan.server.hotrod.logging.Log;
+import org.infinispan.util.KeyValuePair;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.Channel;
+
+@Listener
+class CacheRequestProcessor extends BaseRequestProcessor {
+   private static final Log log = LogFactory.getLog(CacheRequestProcessor.class, Log.class);
+   private static final boolean trace = log.isTraceEnabled();
+   private static final Flag[] LOCAL_NON_BLOCKING_GET = new Flag[] { Flag.CACHE_MODE_LOCAL, Flag.SKIP_CACHE_LOAD };
+   private static final Flag[] SKIP_STATISTICS = new Flag[] { Flag.SKIP_STATISTICS };
+
+   private final HotRodServer server;
+   private final ClientListenerRegistry listenerRegistry;
+   // cacheStopped could be invoked concurrently
+   private final Map<String, CacheInfo> cacheInfo = new ConcurrentHashMap<>();
+
+   CacheRequestProcessor(Channel channel, Executor executor, HotRodServer server) {
+      super(channel, executor);
+      this.server = server;
+      listenerRegistry = server.getClientListenerRegistry();
+      SecurityActions.addListener(server.getCacheManager(), this);
+   }
+
+   @CacheStopped
+   public void cacheStopped(CacheStoppedEvent event) {
+      cacheInfo.remove(event.getCacheName());
+   }
+
+   private CacheInfo getCacheInfo(CacheDecodeContext cdc) {
+      // Fetching persistence manager would require security action, and would be too expensive
+      AdvancedCache<byte[], byte[]> cache = cdc.cache();
+      CacheInfo info = cacheInfo.get(cache.getName());
+      if (info == null) {
+         AdvancedCache<byte[], byte[]> localNonBlocking = cache
+               .noFlags().transform(CacheRequestProcessor::unsetSubject).withFlags(LOCAL_NON_BLOCKING_GET);
+         if (cache.getStatus() != ComponentStatus.RUNNING) {
+            // stay on the safe side
+            return new CacheInfo(localNonBlocking, true, true);
+         }
+         ComponentRegistry cr = SecurityActions.getCacheComponentRegistry(cache);
+         PersistenceManager pm = cr.getComponent(PersistenceManager.class);
+         boolean hasIndexing = SecurityActions.getCacheConfiguration(cache).indexing().index().isEnabled();
+         info = new CacheInfo(localNonBlocking, pm.isEnabled(), hasIndexing);
+         cacheInfo.put(cache.getName(), info);
+      }
+      return info;
+   }
+
+   private static AdvancedCache<byte[], byte[]> unsetSubject(AdvancedCache<byte[], byte[]> cache) {
+      if (cache instanceof SecureCacheImpl) {
+         return new SecureCacheImpl<>(SecurityActions.getUnwrappedCache(cache));
+      } else {
+         return cache;
+      }
+   }
+
+   private boolean isBlockingRead(CacheDecodeContext cdc, CacheInfo info) {
+      return info.persistence && !cdc.decoder.isSkipCacheLoad(cdc.header);
```

Comment:
Why are the read/writes blocking when there's persistence? Is it because we don't have async drivers for all types of persistence that can be attached? Or is it because we don't yet have async API for persistence layer within Infinispan?

Reply:
Tha latter

Reply:
Yes, the persistence API does not support async approach, so even if there was an async driver we could not use it.
There are several layers where we could address it. For simplicity I did the checks here despite it's the least effective. I could also let persistence interceptors do invoke the synchronous API from an executor (if needed*), which would give the non-blocking behaviour to embedded mode as well. But the ideal situation would be an async persistence API.

* We could introduce Flag.HINT_SYNC that would be on for the regular `put()` calls and would skip the delegation to another thread pool when the thread is about to wait anyway.